### ISSUE TITLE
Barebones tqdm in submodule

### ISSUE THREAD

**Issue by CrazyPython** [tqdm/tqdm#255]
add tqdm core for performance freaks. (progress bars often mean performance is needed)

Another related but unrelated note: add extra features using MixIns, so users can create their own tqdm classes with customized features.

**Comment by lrq3000** [comment]
I can add easily and quickly a bare version (please do not call this core, this is reserved to the standard tqdm) using some code we have in perf test, but keep in mind that it will have very few features (no multiple/nested progress bars, no EMA, no bar customization, etc.). It will also be less resilient than standard tqdm (lots of checks that won't be added in bare tqdm).

**Comment by CrazyPython** [comment]
@lrq3000 you can rename other people's issues if you own or have write access to the repo :)

**Comment by lrq3000** [comment]
I know but I preferred to tell you :)

Anyway, I worked on this issue, and in the end such a stripped down tqdm is slower than core tqdm (canonical example: strip dynamic_miniters and you increase the overhead drastically). So a barebone tqdm is still in the works but rather as a minimal example for port developers.

So I'm closing this issue since we likely won't implement it as it was described but repurposed for something else.

About mixins, we'll see later, we already have a modular subclassing system and there's no module yet that could benefit from multiple inheritance of two tqdm submodules.

**Comment by CrazyPython** [comment]
@lrq3000 Hm, what if you just used an old version of tqdm?

**Comment by lrq3000** [comment]
Less features, more buggy and less efficient, hence I dropped this approach. I reused the old approach (function-based with closures) but updated with new features and bugfixes.

**Comment by CrazyPython** [comment]
@lrq3000 Understand. tqdm_bare is slower or faster, and by how much? Also speaking of ports to other languages, why don't you join us on tqdm.cpp?

**Comment by lrq3000** [comment]
It should be faster because there's basically no sanity checking on arguments, less minor features (such as `bar_format`), no class (no object-oriented overhead) and no I/O auto-configuration (console size, environment, etc). So overall fewer opcodes. But it's slower than core tqdm:

```
from tqdm import tbrange
for i in tbrange(int(1e8)):
    pass
100%|#######################|100000000/100000000 [01:20<00:00, 1236089.41it/s]
(80.96664071412025, 'seconds')
for i in tqdm.trange(int(1e8)):
    pass
100%|#######################| 100000000/100000000 [00:19<00:00, 5082075.51it/s]
(19.677953824994802, 'seconds')
```

I didn't dig the reason why it's slower, it's surprising but at the same time core tqdm is heavily optimized already. I guess that either there's a bug with the calculation of the rate or dynamic_miniters, or there's an optimized feature that is missing. Or the non class-based architecture is slower, but that would be surprising considering we lost some speed when we transitioned to a class in core tqdm.

**Comment by CrazyPython** [comment]
@lrq3000 I think I know why. `tqdm.trange` is heavily optimized when loop iterations are trivial. Try it when it does real work* and use a profiler instead of raw timing to compare.

*or choose from the thousands of tqdm real world usages

**Comment by lrq3000** [comment]
@CrazyPython no `tqdm.trange` is optimized for the general case of somewhat constant rate with few fluctuations. It's however working very well with varying flux like internet downloading, but there is an issue with the display update rate but that's an issue of UI (#251) not of optimization. `tqdm` is super optimized and `tqdm.trange` is super optimized for any tight loop.

I did some benchmark, see #258 and discuss further there as this issue is closed since the goal was changed.
